### ISSUE TITLE
Do not try to set @JsonUnwrapped fields if using a @JsonCreator constructor

### ISSUE THREAD

**Issue by victornoel** [FasterXML/jackson-databind#1497]
Hi,

After discussing this matter [on the mailing list](https://groups.google.com/d/topic/jackson-user/rXKO8HZbxBk/discussion), I was redirected to creating an issue to ease some use case of `@JsonUnwrapped`.

Basically, my objective is to avoid using mutable objects and exploit as much as possible constructors to construct objects.

Taking this example:
```java
class Location {
    public final int latitude, longitude;

    public Location(int latitude, int longitude) { ... }
}

class Place {
    public final String name;
    @JsonUnwrapped public final Location location;

    @JsonCreator
    public Place(@JsonProperty("name") String name,
                 @JsonProperty("latitude") int latitude, 
                 @JsonProperty("longitude") int longitude) {
        this.name = name;
        this.location = new Location(latitude, longitude);
    }
}
```

Currently, deserializing a `Place` object fails because Jackson first builds the `Place` object (successfully, including the `location` with the correct values) and then tries to set the `location` field. The latter fails because the `latitude` and `longitude` properties were already used to construct the `Place`, and thus are not used to construct the `Location`.
At the end of the deserialization process, the `location` field is initialised with an object with `null` values for `latitude` and `longitude`!

I would expect Jackson not to try to set the `location` field after a `Place` object was successfully constructed since a constructor is expected to build fully constructed objects?

Apparently (according to the discussion on the mailing list), the following code should have worked:
```java
class Location {
    public final int latitude, longitude;

    public Location(int latitude, int longitude) { ... }
}

class Place {
    public final String name;
    @JsonUnwrapped public final Location location;

    @JsonCreator
    public Place(@JsonProperty("name") String name) {
        this.name = name;
        this.location = null;
    }
}
```

I.e., the `Place` object is partially created, and then Jackson would set the `location` field.
In practice it didn't work, not sure why. But with respect to my proposed solution to the first problem of this issue, I'm not sure it is a so good idea that this second way of doing things work: for me a constructor is meant to fully build an object!

FYI, the only solution I found to work is to have a parameter-less constructor and let Jackson handle everything its way. I feel there is some use case where a constructor would really be needed, for example if the properties must be reworked a bit before being set to the object (keeping in mind I don't want to use setters :).

**Comment by victornoel** [comment]
As Tatu Saloranta noted in the mailing list, an alternative solution to this kind of problem would be to be able to specify that the `@JsonUnwrapped` annotation only applies to serialization.

**Comment by dbolotin** [comment]
Possible solution that I found to make `@JsonUnwrap` affect only serialisation, and be able provide custom `@JsonCreator` is to add `@JsonProperty(access = READ_ONLY)` annotation to the "unwrapped" properties.

**Comment by victornoel** [comment]
@dbolotin so the original example (the one in the issue description) works if you add `@JsonProperty(access = READ_ONLY)` to `location`?

I will try to do some tests, thanks for the idea :)

**Comment by dbolotin** [comment]
Yes, the original example should work as expected. I have very similar situation and `@JsonProperty(access = READ_ONLY)` in my case prevented Jackson from doing it's magic of modifying `final` fields after normal construction with `@JsonCreator`.

**Comment by victornoel** [comment]
Excellent, thanks again!

**Comment by lpandzic** [comment]
@victornoel, are you aware of https://github.com/FasterXML/jackson-modules-java8/tree/master/parameter-names?

**Comment by victornoel** [comment]
@lpandzic I am not actually, thanks :)
But I don't see how it is related to this issue, could you expand? Thanks!

**Comment by lpandzic** [comment]
It's not, sorry for digressing. I was looking for solution to this issue and saw that you were defining @JsonCreator and @JsonProperty which you can avoid with parameter names module.

**Comment by victornoel** [comment]
@lpandzic ah ok, anyway, thanks for the information, it's useful :)

**Comment by slavonic-suomi** [comment]
Hello, any updates on this issue?

I have the similar problem, but with pure java, not kotlin.
There is a little example with migration 2.8 -> 2.9 problem [link](https://github.com/slavonic-suomi/jackson-problems-example/blob/master/src/main/java/com/itsupportme/jackson/example/A.java)

**Comment by phraktle** [comment]
`@JsonUnwrapped` also does not work for deserializing properties in `record` classes, since those are final.

**Comment by phraktle** [comment]
@cowtowncoder can I offer any help in fixing this bug?

**Comment by cowtowncoder** [comment]
@phraktle Help always appreciated!

One practical note: I am leaving on a vacation today, back in 10 days -- so will not be checking for updates. Can sync after that.

**Comment by aowss** [comment]
any suggestions on a workaround to make @JsonUnwrapped work with record classes?
https://github.com/FasterXML/jackson-databind/issues/1497#issuecomment-937714547

**Comment by cowtowncoder** [comment]
To request support for `@JsonUnwrapped` for record classes, please file a separate issue, with a simple example of kind of usage you are thinking of. I think this issue is for slightly different problem.

**Comment by JacksonJang** [comment]
@cowtowncoder 
I’ll work on resolving this issue as soon as possible.

**Comment by cowtowncoder** [comment]
@JacksonJang I think this requires `DeserializationFeature` or something similar to #1709.

And/or possibly new property for `@JsonUnwrapped`, `OptBoolean enabledForDeserialization();` (defaults to... `OptBoolean.TRUE`?)
... or should that instead be `OptBoolean serializationOnly();` ?

Overall we need to figure exact desired logic before trying to implement a solution. This is tricky.

I almost wonder if re-surrecting PR #5500 is the way to go, after all.
That is, not creating those Unwrapped POJO values for which no properties were buffered (and the new `DeserializationFeature` is enabled).

**Comment by JacksonJang** [comment]
@cowtowncoder 
In my view, for the case discussed in https://github.com/FasterXML/jackson-databind/pull/5500, adding a new option as a `DeserializationFeature` seems to be a better direction.

For this issue, I think it would be preferable to handle it as follows:
```
@JsonUnwrapped(deserialize = true)
@JsonUnwrapped(deserialize = false)
or
@JsonUnwrapped(serialize = true)
@JsonUnwrapped(serialize = false)
```

I’d appreciate your thoughts on whether this approach makes sense.

**Comment by cowtowncoder** [comment]
I think either approach would work for both cases.

But thinking through complexity, adding and using a `DeserializationFeature` is much easier to implement, although less granular (affects all usage).

On annotations; plain "serialize" or "deserialize" is not very descriptive. But let's not worry about that yet -- ignore my earlier comment wr #5500 requiring annotation. Let's add `DeserializationFeature` like you suggested, but maybe name as `USE_NULL_FOR_EMPTY_UNWRAPPED`

**Comment by JacksonJang** [comment]
That makes sense to me.
I agree that introducing a `DeserializationFeature` is the simpler and more practical approach overall.
